sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/enhancedfanout/ShardSubscribersPoolImpl.java

```diff
@@ -0,0 +1,404 @@
+    // FIXME: If future completes with unrecoverable error, entire reader should fail
+    // FIXME: Down-sharding can cause shards to be "lost"! Re-visit this logic!
+    LOG.info("Processing re-shard signal {}", event);
+
+    List<String> successorShardsIds = new ArrayList<>();
+
+    for (ChildShard childShard : event.getChildShards()) {
+      if (childShard.parentShards().contains(event.getShardId())) {
+        if (childShard.parentShards().size() > 1) {
+          // This is the case of merging two shards into one.
+          // when there are 2 parent shards, we only pick it up if
+          // its max shard equals to sender shard ID
+          String maxId = childShard.parentShards().stream().max(String::compareTo).get();
+          if (event.getShardId().equals(maxId)) {
+            successorShardsIds.add(childShard.shardId());
+          }
+        } else {
+          // This is the case when shard is split
+          successorShardsIds.add(childShard.shardId());
+        }
+      }
+    }
+
+    if (successorShardsIds.isEmpty()) {
+      LOG.info("Found no successors for shard {}", event.getShardId());
+    } else {
+      LOG.info("Found successors for shard {}: {}", event.getShardId(), successorShardsIds);
+    }
+    return successorShardsIds;
+  }
+
+  private CompletableFuture<Void> createSubscription(
+      String shardId, ShardCheckpoint checkpoint, boolean isInitial) {
+    UUID subscribeRequestId = UUID.randomUUID();
+    SubscribeToShardRequest request =
+        SubscribeToShardRequest.builder()
+            .consumerARN(config.getConsumerArn())
+            .shardId(shardId)
+            .startingPosition(checkpoint.toStartingPosition())
+            .build();
+
+    LOG.info("Pool id {} starting subscribe request {} - {}", poolId, subscribeRequestId, request);
+
+    CountDownLatch eventsHandlerReadyLatch = new CountDownLatch(1);
+
+    KinesisShardEventsSubscriber shardEventsSubscriber =
+        new KinesisShardEventsSubscriber(
+            this,
+            eventsHandlerReadyLatch,
+            config.getStreamName(),
+            config.getConsumerArn(),
+            shardId);
+
+    SubscribeToShardResponseHandler responseHandler =
+        SubscribeToShardResponseHandler.builder()
+            .onError(
+                e ->
+                    LOG.error(
+                        "Pool id = {} failed to execute subscribe request {} - {}",
+                        poolId,
+                        subscribeRequestId,
+                        request,
+                        e))
+            .subscriber(() -> shardEventsSubscriber)
+            .build();
+
+    CompletableFuture<Void> f = kinesis.subscribeToShard(request, responseHandler);
+    ShardSubscriberState shardState =
+        new ShardSubscriberStateImpl(f, shardEventsSubscriber, checkpoint);
+    boolean subscriptionWasEstablished = false;
+    try {
+      subscriptionWasEstablished =
```

Comment:
Waiting here for the latch makes `createSubscription ` blocking, despite returning a `CompletableFuture` - which makes the caller belief this is actually async and non blocking. This is something to always watch out for.

In this cases it's simple, we can rely on timeouts in the underlaying AWS SDK. If we fail to subscribe, the future returned by the AWS client will eventually fail. That's all that matters, and by then we can react to it using something like
`kinesis.subscribeToShard(request, responseHandler).whenCompleteAsync(resubscriptionHandler)`.

Note, we also don't want to return the future returned by Kinesis. But instead, another future we can control ourselves to:
- complete it with a `Void` value (`null`) when we've consumed the entire shard and the shard isn't active anymore or if the subscriber was stopped.
- complete it with an exception if a non retryable error occured .

That allows to hide the re-subscription from the outside caller. All the pool should care about is if the subscriber is active or if it succeeded or failed.